File: 01_Array_Hashings/04_Group_Anagrams/49_Group_Anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> res;

        for (string s: strs) {
            int arr[26] = {};
            for (char c: s) {
                arr[int(c) - int('a')] += 1;
            }
            string key;
            for (int v : arr){
                key += to_string(v) + '#';
            }
            res[key].push_back(s);
        };

        vector<vector<string>> result;
        for (auto& pair : res){
            result.push_back(pair.second);
        };

        return result;
    }
};

File: 01_Array_Hashings/04_Group_Anagrams/49_Group_Anagrams.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> res = new HashMap<>();

        for (String s : strs) {
            int [] arr = new int[26];
            Arrays.fill(arr, 0);

            char [] s_arr = s.toCharArray();
            for (char c: s_arr){
                arr[c-'a'] += 1;
            }

            String key = Arrays.toString(arr);

            if (!res.containsKey(key)){
                res.put(key, new ArrayList<>());
            }
            res.get(key).add(s);
        };
        return new ArrayList<>(res.values());
    }
}